src/methods/clipping/intersection.jl

```diff
@@ -0,0 +1,423 @@
+    intersection(
+        ::GI.LineTrait, line_a,
+        ::GI.LineTrait, line_b,
+    )::Union{
+        ::Tuple{::Real, ::Real},
+        ::Nothing
+    }
+
+Calculates the intersection between two line segments. Return nothing if
+there isn't one.
+"""
+function intersection(::GI.LineTrait, line_a, ::GI.LineTrait, line_b)
+    # Get start and end points for both lines
+    a1 = GI.getpoint(line_a, 1)
+    a2 = GI.getpoint(line_a, 2)
+    b1 = GI.getpoint(line_b, 1)
+    b2 = GI.getpoint(line_b, 2)
+    # Determine the intersection point
+    point, fracs = _intersection_point((a1, a2), (b1, b2))
+    # Determine if intersection point is on line segments
+    if !isnothing(point) && 0 <= fracs[1] <= 1 && 0 <= fracs[2] <= 1
+        return point
+    end
+    return nothing
+end
+
+intersection(
+    trait_a::Union{GI.LineStringTrait, GI.LinearRingTrait},
+    geom_a,
+    trait_b::Union{GI.LineStringTrait, GI.LinearRingTrait},
+    geom_b,
+) = intersection_points(trait_a, geom_a, trait_b, geom_b)
+
+"""
+    intersection(
+        ::GI.AbstractTrait, geom_a,
+        ::GI.AbstractTrait, geom_b,
+    )::Union{
+        ::Vector{Vector{Tuple{::Real, ::Real}}}, # is this a good return type?
+        ::Nothing
+    }
+
+Calculates the intersection between two line segments. Return nothing if
+there isn't one.
+"""
+function intersection(
+    trait_a::GI.AbstractTrait, geom_a,
+    trait_b::GI.AbstractTrait, geom_b,
+)
+    @assert(
+        false,
+        "Intersection between $trait_a and $trait_b isn't implemented yet.",
+    )
+    return nothing
+end
+
+"""
+    _trace_intersection(poly_a, poly_b, a_list, b_list, a_idx_list,
+      intr_list, edges_a, edges_b)::Vector{Vector{Tuple{Float64}}}
+
+Traces the outlines of two polygons in order to find their intersection.
+It returns the outlines of all polygons formed in the intersection. If
+they do not intersect, it returns an empty array.
+
+"""
+function _trace_intersection(poly_a, poly_b, a_list, b_list, a_idx_list, intr_list, edges_a, edges_b)
+    # Pre-allocate array for return polygons
+    return_polys = Vector{Vector{Tuple{Float64, Float64}}}(undef, 0)
+
+    # Keep track of number of processed intersection points
+    processed_pts = 0
+    tracker = copy(a_idx_list)
+
+    while processed_pts < length(intr_list)
+        # Create variables "list_edges" and "list" so that we can toggle between
+        # a_list and b_list
+        list_edges = edges_a
+        list = a_list
+
+        # Find index of first unprocessed intersecting point in subject polygon
+        starting_pt = minimum(tracker)
+        idx = starting_pt
+
+        # Get current first unprocessed intersection point PolyNode
+        current = a_list[idx]
+        # Initialize array to store the intersection polygon cartesian points
+        pt_list = Vector{Tuple{Float64, Float64}}(undef, 0)
+        # Add the first point to the array
+        push!(pt_list, (intr_list[current.idx][1], intr_list[current.idx][2]))
```

Comment:
This will be a lot cleaner whenw e can just take the point from the PolyNode and push the point

Reply:
Also, can we move the first point analysis into the below while loop, since that code is repeated down below within the while loop? And then move the index update to the bottom of the while loop